packages/template-retail-react-app/app/contexts/index.js

```diff
@@ -27,23 +29,162 @@ import PropTypes from 'prop-types'
  *
  * import {useCategories} from './hooks'
  *
- * const {categories, setCategories} = useCategories()
+ * const {root, findAndModifyFirst, itemsKey, fetchCategoryNode} = useCategories()
  *
  */
 export const CategoriesContext = React.createContext()
-export const CategoriesProvider = ({categories: initialCategories = {}, children}) => {
-    const [categories, setCategories] = useState(initialCategories)
+export const CategoriesProvider = ({treeRoot = {}, children, locale}) => {
+    const itemsKey = 'categories'
+    const DEFAULT_ROOT_CATEGORY = 'root'
+    const LOCAL_STORAGE_PREFIX = `pwa-kit-cat-`
+
+    const api = useCommerceAPI()
+    const [root, setRoot] = useState({
+        // map over the server-provided cat
+        ...treeRoot[DEFAULT_ROOT_CATEGORY],
+        [itemsKey]: treeRoot[DEFAULT_ROOT_CATEGORY]?.[itemsKey]?.map((item) => ({
+            ...item,
+            loaded: item?.loaded ?? false
+        }))
+    })
+
+    // iterates through each deep nested object and if finds object that has prop and value specified in objToFindBy
+    // argument, it replaces the current object with replacementObj, stops recursive walk and returns the whole tree.
+    // If none is found, it returns false.
+    const findAndModifyFirst = (tree, childrenKey, objToFindBy, replacementObj) => {
```

Comment:
As a general best practices, a function should not mutate its input. Mutating the input leaks to unexpected issues and makes code harder to understand.

```js
    const findAndModifyFirst = (tree, childrenKey, objToFindBy, replacementObj) => {
            const clone = Object.assign({}, tree)
            // do work ...
            return clone
   }
```

Reply:
Good suggestion. @wjhsf had mentioned this and I agree that we need to not mutate the input

Reply:
Actually @kevinxh my preference (not a big deal) is to use object spread which handles cloning implicitly.

```javascript
    const findAndModifyFirst = (tree, childrenKey, objToFindBy, replacementObj) => {
            // do work 
            return {
              ...tree
            }
   }
```